# Embedding Model using _sentance_transformer_

In this notebook, we explore the way that text with multiple words/tokens is encoded into an embedding vector using the popular [sentence_transformers library](https://sbert.net/index.html). We will Check:

* [OpenAI Embedding](#openai-embedding)
* [Open source encoder input embeddings](#open-source-encoder---input-embeddings)
* [Open source encoder output embedding (with context)](#open-source-encoder---output-embedding-with-context)
* [Improved encoder for queries and documents (bi-encoder)](#improved-encoder-for-queries-and-documents-bi-encoder)

Define rich theme for better object printing

In [1]:
from rich.console import Console
from rich_theme_manager import Theme, ThemeManager
import pathlib

theme_dir = pathlib.Path("themes")
theme_manager = ThemeManager(theme_dir=theme_dir)
dark = theme_manager.get("dark")

# Create a console with the dark theme
console = Console(theme=dark)


In [2]:
import warnings

# Suppress warnings
warnings.filterwarnings('ignore')

## OpenAI Embedding

A common option is to use the embedding from the same provider as the generation model.

In [3]:
first_sentence = "I have no interest in politics"

In [5]:
import sys
!{sys.executable} -m pip install python-dotenv

  Using cached python_dotenv-1.2.1-py3-none-any.whl.metadata (25 kB)
Using cached python_dotenv-1.2.1-py3-none-any.whl (21 kB)


In [6]:
from dotenv import load_dotenv

load_dotenv()

False

In [7]:
from sentence_transformers import SentenceTransformer

embedding_model = SentenceTransformer('all-MiniLM-L6-v2')
embedding = embedding_model.encode(first_sentence)

console.print(embedding)
console.print(f"Embedding dimension: {len(embedding)}")

The cache for model files in Transformers v4.22.0 has been updated. Migrating your old cache. This is a one-time only operation. You can interrupt this and resume the migration later on by calling `transformers.utils.move_cache()`.


0it [00:00, ?it/s]

[ 8.04471150e-02 -4.18708846e-02 -6.72248797e-03  6.22915477e-03
  5.51091954e-02 -5.64353075e-03  1.93882044e-02  6.39811810e-03
 -3.26935910e-02  2.66913138e-03 -6.59479350e-02 -7.77595416e-02
 -3.95389199e-02 -2.73962189e-02 -1.30916759e-02  3.70858870e-02
 -7.08684996e-02 -3.59889008e-02  5.01063513e-03  3.44360396e-02
 -3.26007046e-03  1.24481479e-02  2.39544716e-02 -4.10691611e-02
  6.47743931e-03 -3.65413638e-04  2.04820884e-03 -7.95463920e-02
 -4.74378318e-02 -6.30370080e-02  5.88013828e-02  2.97880620e-02
 -1.70271304e-02  5.01092039e-02  1.32132664e-01  1.49379845e-03
  5.93851097e-02  3.27357315e-02  9.03991833e-02  2.34288052e-02
  4.30272557e-02 -1.33918775e-02  3.66541115e-03 -3.28965262e-02
  3.08137313e-02  3.17489952e-02 -2.27517057e-02 -5.22434935e-02
  3.65324132e-02 -1.18004084e-02  1.78214945e-02  3.74964997e-02
 -5.16330153e-02  4.31708135e-02  8.41141641e-02 -9.52073336e-02
 -4.43788096e-02  8.74651596e-02 -8.18211120e-03 -7.39176050e-02
  2.12724339e-02 -1.51653048e-02 -2.47709658e-02 -2.95977574e-02
 -5.53318858e-02  3.84192131e-02 -3.26747969e-02  8.49790052e-02
  4.58539277e-03 -1.90317247e-03 -2.76655667e-02  3.12124062e-02
 -5.98950945e-02  1.30132642e-02 -6.34549744e-03 -9.48312357e-02
  5.81262596e-02  1.12949520e-01  1.49267703e-01  5.27240448e-02
 -9.77183878e-03 -1.91670563e-02 -1.01165660e-01 -3.72959510e-03
 -4.22407463e-02 -1.43098296e-03 -8.79553240e-03  3.43777947e-02
  1.60052739e-02  9.47199413e-04 -4.01494317e-02  6.70321807e-02
  3.88570689e-02  1.02219600e-02 -1.77685115e-02  6.68197684e-03
  4.94260117e-02  7.18135536e-02 -1.04377158e-01  1.37055129e-01
 -4.70148325e-02 -3.40478718e-02 -9.52642411e-02  5.64329401e-02
 -2.21229903e-02  6.92287907e-02 -8.27498809e-02  5.09969257e-02
 -4.08956707e-02 -7.16010248e-03 -1.06219416e-02  6.04808971e-04
  4.69893310e-03  5.76898977e-02 -1.56199839e-02  1.96696334e-02
  2.41343193e-02 -2.29868684e-02  5.69786914e-02  4.72150743e-02
 -4.43495512e-02 -3.08970250e-02 -8.97128955e-02  4.99526598e-02
  2.35068798e-02 -7.83377215e-02 -3.52642201e-02 -2.88514987e-33
 -2.12550666e-02 -2.92803384e-02  1.76795162e-02  1.78022776e-02
 -1.79856904e-02  9.77589116e-02 -6.18543942e-03 -5.06699905e-02
 -9.70208645e-03  2.39158478e-02  4.88723256e-02  1.11008488e-01
  3.75732332e-02 -1.07060820e-01  5.06583564e-02 -6.46521477e-03
 -2.46913936e-02 -6.52062707e-03  4.37063761e-02  2.65525505e-02
 -5.00257686e-02  3.80925871e-02 -1.43195176e-02 -1.21446615e-02
  1.67451650e-02 -2.34878119e-02  9.25623551e-02 -9.16953087e-02
  2.90531460e-02  3.04557011e-02 -1.27393827e-02  5.58118001e-02
 -3.47622596e-02 -8.27111751e-02 -1.58141050e-02  1.93751492e-02
 -6.80302605e-02 -3.01776417e-02 -9.73517448e-03 -3.27085033e-02
 -2.64319684e-02  1.92714985e-02 -3.33821699e-02  2.59442464e-03
  2.91614272e-02 -1.48186125e-02  1.41925752e-01 -5.18426225e-02
 -7.27149248e-02  1.86627228e-02  4.44305725e-02 -2.55811773e-02
 -2.33109966e-02 -2.17366200e-02 -8.16528276e-02 -7.33707249e-02
 -3.95029522e-02 -1.49882548e-02 -4.94570658e-02 -1.07098304e-01
  3.71844992e-02 -4.14417796e-02 -5.35639934e-02  3.19113992e-02
 -5.38006984e-02  1.18423747e-02 -9.24026668e-02 -2.38173604e-02
  3.66845801e-02 -6.98151737e-02  8.72825098e-04 -1.29450615e-02
 -9.49374679e-03 -2.34678108e-02 -7.73386359e-02  4.40445207e-02
 -7.02998508e-03 -4.61321548e-02 -2.59958412e-02  4.60969992e-02
  6.60139620e-02 -2.37852205e-02  3.54572833e-02  6.85421005e-03
  9.47870091e-02 -8.81897844e-03  7.34627098e-02 -1.48385996e-02
  1.05371095e-01 -4.97624576e-02 -1.31978169e-01 -4.58532311e-02
 -3.65228355e-02  5.52779203e-03 -2.45569646e-02  1.51629992e-33
 -2.88547073e-02 -6.18994944e-02  3.36548090e-02  2.84650382e-02
  1.20678796e-02  5.22785336e-02  9.14553478e-02 -4.55580913e-02
  9.53655168e-02  3.65554094e-02  2.72871070e-02 -2.25056894e-03
  5.67435697e-02  1.05469771e-01  8.69187191e-02 -1.88994706e-02
  4.25998345e-02  2.73355022e-02 -1.18605599e-01  6.85793087e-02
 -5.32611161e-02  1.22478783e-01 -6

Embedding dimension: 384

## Open source encoder - input embeddings

We will start with one of the popular encoders from the _sentence_transformers_ library. It will allow us to explore its architecture and flow, and later on to optimize it to our use case.

In [8]:
from sentence_transformers import SentenceTransformer

model = SentenceTransformer("all-MiniLM-L6-v2")

### Model Tokenizer

We will use the default tokenizer of the model. Every word or subword is converted into a token with a constant ID. For example, in the following two sentences, the word `interest` is tokenized to the same ID (`3037`).

In [9]:
first_sentence = "I have no interest in politics"
second_sentence = "The bank's interest rate rises"


In [10]:
tokenized_first_sentence = model.tokenize([first_sentence])
console.rule(f"{first_sentence}")
console.print(tokenized_first_sentence)

───────────────────────────────────────── I have no interest in politics ──────────────────────────────────────────

{
    'input_ids': tensor([[ 101, 1045, 2031, 2053, 3037, 1999, 4331,  102]]),
    'token_type_ids': tensor([[0, 0, 0, 0, 0, 0, 0, 0]]),
    'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 1]])
}

In [11]:
tokenized_second_sentence = model.tokenize([second_sentence])
console.rule(f"{second_sentence}")
console.print(tokenized_second_sentence)

───────────────────────────────────────── The bank's interest rate rises ──────────────────────────────────────────

{
    'input_ids': tensor([[ 101, 1996, 2924, 1005, 1055, 3037, 3446, 9466,  102]]),
    'token_type_ids': tensor([[0, 0, 0, 0, 0, 0, 0, 0, 0]]),
    'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1]])
}

The token ID can be used to convert it back into readable text:

In [12]:
sentence_tokens = (
    model
    .tokenizer
    .convert_ids_to_tokens(
        tokenized_second_sentence["input_ids"]
        [0]
    )
)

console.print(sentence_tokens)

['[CLS]', 'the', 'bank', "'", 's', 'interest', 'rate', 'rises', '[SEP]']

### Model Vocabulary

In [13]:
vocabulary = (
    model
    ._first_module()
    .tokenizer
    .get_vocab()
    .items()
)

console.print("[bold]Vocabulary size[/bold]:", len(vocabulary))
console.print(dict(list(vocabulary)[:20]))

Vocabulary size: 30522

{
    '♣': 1624,
    'slopes': 10314,
    'organ': 5812,
    '##kos': 15710,
    'convertible': 22840,
    'breeds': 15910,
    'tinted': 25577,
    '÷': 1099,
    'ceramic': 14692,
    'gerry': 14926,
    'cause': 3426,
    'excavations': 14018,
    'brianna': 25558,
    '1634': 28502,
    'portland': 6734,
    '[SEP]': 102,
    'boys': 3337,
    'commencement': 20561,
    'churning': 26765,
    '##uld': 21285
}

Let's see part of the tokenizer vocabulary. We will search for the token for `interest` and see its neighbors.


In [14]:


sorted_vocabulary = sorted(
    vocabulary, 
    key=lambda x: x[1],  # uses the value of the dictionary entry
)
sorted_tokens = [token for token, _ in sorted_vocabulary]

focused_token = 'interest'
# Find the index of the 'interest' token
focused_index = sorted_tokens.index(focused_token)

# Get 20 tokens around the focused token
start_index = max(0, focused_index - 10)
end_index = min(len(sorted_tokens), focused_index + 11)
tokens_around_focused_index = sorted_tokens[start_index:end_index]

from rich.table import Table

table = Table(title=f"Tokens around '{focused_token}':")
table.add_column("id", justify="right", style="cyan", no_wrap=True)
table.add_column("token", style="bright_green")

for i, token in enumerate(tokens_around_focused_index, start=start_index):
    if token == focused_token:
        table.add_row(f"[bold][black on yellow]{i}[/black on yellow][/bold]", f"[bold][black on yellow]{token}[/black on yellow][/bold]")
    else:
        table.add_row(str(i), token)

console.print(table)

     Tokens around     
      'interest':      
┏━━━━━━┳━━━━━━━━━━━━━━┓
┃   id ┃ token        ┃
┡━━━━━━╇━━━━━━━━━━━━━━┩
│ 3027 │ ft           │
│ 3028 │ valley       │
│ 3029 │ organization │
│ 3030 │ stopped      │
│ 3031 │ onto         │
│ 3032 │ countries    │
│ 3033 │ parts        │
│ 3034 │ conference   │
│ 3035 │ queen        │
│ 3036 │ security     │
│ 3037 │ interest     │
│ 3038 │ saying       │
│ 3039 │ allowed      │
│ 3040 │ master       │
│ 3041 │ earlier      │
│ 3042 │ phone        │
│ 3043 │ matter       │
│ 3044 │ smith        │
│ 3045 │ winning      │
│ 3046 │ try          │
│ 3047 │ happened     │
└──────┴──────────────┘

### The Embedding Transformer Model 

Transformer consists of multiple stack modules. Tokens are an input of the first one. Let's see that first model.

In [15]:
console.print(model)

SentenceTransformer(
  (0): Transformer({'max_seq_length': 256, 'do_lower_case': False, 'architecture': 'BertModel'})
  (1): Pooling({'word_embedding_dimension': 384, 'pooling_mode_cls_token': False, 'pooling_mode_mean_tokens': True,
'pooling_mode_max_tokens': False, 'pooling_mode_mean_sqrt_len_tokens': False, 'pooling_mode_weightedmean_tokens': 
False, 'pooling_mode_lasttoken': False, 'include_prompt': True})
  (2): Normalize()
)

In [16]:
first_module = model._first_module()
console.print(first_module.auto_model)

BertModel(
  (embeddings): BertEmbeddings(
    (word_embeddings): Embedding(30522, 384, padding_idx=0)
    (position_embeddings): Embedding(512, 384)
    (token_type_embeddings): Embedding(2, 384)
    (LayerNorm): LayerNorm((384,), eps=1e-12, elementwise_affine=True)
    (dropout): Dropout(p=0.1, inplace=False)
  )
  (encoder): BertEncoder(
    (layer): ModuleList(
      (0-5): 6 x BertLayer(
        (attention): BertAttention(
          (self): BertSdpaSelfAttention(
            (query): Linear(in_features=384, out_features=384, bias=True)
            (key): Linear(in_features=384, out_features=384, bias=True)
            (value): Linear(in_features=384, out_features=384, bias=True)
            (dropout): Dropout(p=0.1, inplace=False)
          )
          (output): BertSelfOutput(
            (dense): Linear(in_features=384, out_features=384, bias=True)
            (LayerNorm): LayerNorm((384,), eps=1e-12, elementwise_affine=True)
            (dropout): Dropout(p=0.1, inplace=False)
          )
        )
        (intermediate): BertIntermediate(
          (dense): Linear(in_features=384, out_features=1536, bias=True)
          (intermediate_act_fn): GELUActivation()
        )
        (output): BertOutput(
          (dense): Linear(in_features=1536, out_features=384, bias=True)
          (LayerNorm): LayerNorm((384,), eps=1e-12, elementwise_affine=True)
          (dropout): Dropout(p=0.1, inplace=False)
        )
      )
    )
  )
  (pooler): BertPooler(
    (dense): Linear(in_features=384, out_features=384, bias=True)
    (activation): Tanh()
  )
)

and focus on the `embeddings` part:

In [17]:
embeddings = first_module.auto_model.embeddings
console.print(embeddings)

BertEmbeddings(
  (word_embeddings): Embedding(30522, 384, padding_idx=0)
  (position_embeddings): Embedding(512, 384)
  (token_type_embeddings): Embedding(2, 384)
  (LayerNorm): LayerNorm((384,), eps=1e-12, elementwise_affine=True)
  (dropout): Dropout(p=0.1, inplace=False)
)

### Embedding Model Input Token IDs

We will send to the transformer model the two sentences above, and check the embedding similarity between the **input** tokens.

In [18]:
import torch

device = torch.device("mps" if torch.has_mps else "cpu")  # Use MPS for Apple, CUDA for others, or fallback to CPU

with torch.no_grad():
    # Tokenize both texts
    first_tokens = model.tokenize([first_sentence])
    second_tokens = model.tokenize([second_sentence])
    
    # Get the corresponding embeddings
    first_embeddings = embeddings.word_embeddings(
        first_tokens["input_ids"].to(device)
    )
    second_embeddings = embeddings.word_embeddings(
        second_tokens["input_ids"].to(device)
    )

console.print(first_embeddings.shape, second_embeddings.shape)

torch.Size([1, 8, 384])
torch.Size([1, 9, 384])

In [19]:
from rich.table import Table

table = Table(title="Embeddings Shape Explanation")

table.add_column("Text", style="cyan", no_wrap=True)
table.add_column("Batch Size", style="white")
table.add_column("Tokens Number", style="white")
table.add_column("Embedding Dimension", style="white")

table.add_row(
    first_sentence,
    str(first_embeddings.shape[0]),
    str(first_embeddings.shape[1]),
    str(first_embeddings.shape[2]),
)
table.add_row(
    second_sentence,
    str(second_embeddings.shape[0]),
    str(second_embeddings.shape[1]),
    str(second_embeddings.shape[2]),
)

console.print(table)

                            Embeddings Shape Explanation                             
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━┓
┃ Text                           ┃ Batch Size ┃ Tokens Number ┃ Embedding Dimension ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━┩
│ I have no interest in politics │ 1          │ 8             │ 384                 │
│ The bank's interest rate rises │ 1          │ 9             │ 384                 │
└────────────────────────────────┴────────────┴───────────────┴─────────────────────┘

### Compare the input embedding of the tokens

In [20]:
from sentence_transformers import util
import altair as alt
import pandas as pd

# Calculate cosine similarity
distances = util.cos_sim(
    first_embeddings.squeeze(), 
    second_embeddings.squeeze()
).cpu().numpy()

# Get token labels
x_labels = model.tokenizer.convert_ids_to_tokens(second_tokens["input_ids"][0])
y_labels = model.tokenizer.convert_ids_to_tokens(first_tokens["input_ids"][0])

# Create a DataFrame for Altair
data = pd.DataFrame(
    [(x, y, distances[i, j]) for i, y in enumerate(y_labels) for j, x in enumerate(x_labels)],
    columns=['x', 'y', 'similarity']
)

# Create heatmap using Altair
chart = alt.Chart(data).mark_rect().encode(
    x=alt.X('x:O', title='Second Sentence Tokens', axis=alt.Axis(labelAngle=-45), sort=x_labels),
    y=alt.Y('y:O', title='First Sentence Tokens', sort=y_labels),
    color=alt.Color('similarity:Q', scale=alt.Scale(scheme='yellowgreenblue')),
    tooltip=['x', 'y', alt.Tooltip('similarity:Q', format='.2f')]
).properties(
    width=500,
    height=400,
    title='Input Token Similarity Heatmap'
)

# Add text labels
text = chart.mark_text(baseline='middle').encode(
    text=alt.Text('similarity:Q', format='.2f'),
    color=alt.condition(
        alt.datum.similarity > 0.5,
        alt.value('white'),
        alt.value('black')
    )
)

# Combine chart and text
final_chart = (chart + text).configure_title(fontSize=16)

# Display the chart
final_chart

alt.LayerChart(...)

### Vocabulary Embedding

As we saw there are 30,522 tokens in the vocabulary, and each token is embedded with a vector of size 384.

In [ ]:
token_embeddings = first_module.auto_model \
    .embeddings \
    .word_embeddings \
    .weight \
    .detach() \
    .cpu() \
    .numpy()

console.print(token_embeddings.shape)

### Reduce the embedding vectors to 2D for visualization

We will use the TSNE library to create a 2D visualization of the token embeddings, to allow us to see tokens that are close to one another.

This process can take about a minute or two, based on your CPU.

In [ ]:
from sklearn.manifold import TSNE

tsne = TSNE(n_components=2, metric="cosine", random_state=42)
tsne_embeddings_2d = tsne.fit_transform(token_embeddings)
console.print(tsne_embeddings_2d.shape)

### Token Embedding Visualization

Once we have the 384 dimension reduced to 2D, we can plot it to explore it.

In [ ]:
token_colors = []
for token in sorted_tokens:
    if token[0] == "[" and token[-1] == "]": # Control Tokens
        token_colors.append("red")
    elif token.startswith("##"):            # Suffix Tokens
        token_colors.append("blue")
    else:
        token_colors.append("green")        # All Word Tokens

In [ ]:
import altair as alt
import pandas as pd

# Disable max rows limit
alt.data_transformers.disable_max_rows()

df = pd.DataFrame({
    'x': tsne_embeddings_2d[:, 0],
    'y': tsne_embeddings_2d[:, 1],
    'token': sorted_tokens,
    'color': token_colors
})

chart = alt.Chart(df).mark_circle(size=30).encode(
    x='x:Q',
    y='y:Q',
    color=alt.Color('color:N', scale=None),
    tooltip=['token:N']
).properties(
    width=600,
    height=900,
    title='Token Embeddings'
).interactive()

chart

## Open source encoder - output embedding (with context)

Now let's see the token embeddings at the output of the transformer embedding model.

In [ ]:
output_embedding = model.encode([first_sentence])
console.print(output_embedding.shape)

In [ ]:
output_token_embeddings = model.encode(
    [first_sentence], 
    output_value="token_embeddings"
)
console.print(output_token_embeddings[0].shape)

In [ ]:
with torch.no_grad():
    first_tokens = model.tokenize([first_sentence])
    second_tokens = model.tokenize([second_sentence])
    
    first_output_embeddings = model.encode(
        [first_sentence], 
        output_value="token_embeddings"
    )
    second_output_embeddings = model.encode(
        [second_sentence], 
        output_value="token_embeddings"
    )

# Calculate cosine similarity
distances = util.cos_sim(
    first_output_embeddings[0], 
    second_output_embeddings[0]
)

### Visualize the **Output** Token Similarity

Similar to the way to visualized the similarity of the **Input** tokens from the embedding lookup table, we will visualize the similar of the same tokens from the **Output** of the transformer model, after it applied the positional encoding and the attention layers.

In [ ]:
import numpy as np

# Convert tensor to numpy if needed
if hasattr(distances, 'numpy'):
    distances_np = distances.detach().numpy()
else:
    distances_np = distances

# Get token labels
x_labels = model.tokenizer.convert_ids_to_tokens(second_tokens["input_ids"][0])
y_labels = model.tokenizer.convert_ids_to_tokens(first_tokens["input_ids"][0])

# Create a DataFrame for Altair
data = pd.DataFrame(
    [(x, y, float(distances_np[i, j])) for i, y in enumerate(y_labels) for j, x in enumerate(x_labels)],
    columns=['x', 'y', 'similarity']
)

# Create heatmap using Altair
chart = alt.Chart(data).mark_rect().encode(
    x=alt.X('x:O', title='Second Sentence Tokens', axis=alt.Axis(labelAngle=-45), sort=x_labels),
    y=alt.Y('y:O', title='First Sentence Tokens', sort=y_labels),
    color=alt.Color('similarity:Q', scale=alt.Scale(scheme='yellowgreenblue', domain=[0, 1])),
    tooltip=['x', 'y', alt.Tooltip('similarity:Q', format='.2f')]
).properties(
    width=500,
    height=400,
    title='Output Token Similarity Heatmap'
)

# Add text labels
text = chart.mark_text(baseline='middle').encode(
    text=alt.Text('similarity:Q', format='.2f'),
    color=alt.condition(
        alt.datum.similarity > 0.5,
        alt.value('white'),
        alt.value('black')
    )
)

# Combine chart and text
final_chart = (chart + text).configure_title(fontSize=16)
final_chart


In [ ]:
# Calculate cosine distance between output embeddings
from sklearn.metrics.pairwise import cosine_distances
from rich.panel import Panel
from rich.table import Table

def calculate_sentence_similarity(first_sentence, second_sentence):

    first_embeddings = model.encode([first_sentence])
    second_embeddings = model.encode([second_sentence])

    # Reshape the embeddings to 2D arrays
    first_embedding_2d = first_embeddings.reshape(1, -1)
    second_embedding_2d = second_embeddings.reshape(1, -1)

    # Calculate cosine distance
    cosine_distance = cosine_distances(first_embedding_2d, second_embedding_2d)[0][0]

    # Note: Cosine distance is 1 - cosine similarity
    cosine_similarity = 1 - cosine_distance

    console.print(
        Panel(
            f"[cyan bold]First Sentence:[/cyan bold] {first_sentence}\n"
            f"[cyan bold]Second Sentence:[/cyan bold] {second_sentence}",
            title="[green bold]Similarity Calculation[/green bold]",
            expand=False,
            border_style="dim white"
        )
    )

    results = Table(title="Results")
    results.add_column("Metric", style="bold")
    results.add_column("Value", style="bold")
    results.add_row("Cosine Distance", f"{cosine_distance:.4f}", style="cyan")
    results.add_row("Cosine Similarity", f"{cosine_similarity:.4f}", style="bright_yellow")

    console.print(results)


In [ ]:
calculate_sentence_similarity(first_sentence, second_sentence)

In [ ]:
third_sentence = "Chase increased its lending fees"

calculate_sentence_similarity(second_sentence, third_sentence)

## Improved encoder for queries and documents (bi-encoder)

We will use [Contextual Document Embeddings (CDE)](https://huggingface.co/jxm/cde-small-v1), which is one of the trending models in the Hugging Face model repository.

In [ ]:
import sys
!{sys.executable} -m pip install einops

In [ ]:
import sys
!{sys.executable} -m pip install "transformers==4.44.0"

In [ ]:
import transformers

improved_model = transformers.AutoModel.from_pretrained("jxm/cde-small-v1", trust_remote_code=True)
tokenizer = transformers.AutoTokenizer.from_pretrained("bert-base-uncased")

In [ ]:
console.print(improved_model)

In [ ]:
from datasets import load_dataset

corpus = load_dataset("BeIR/fiqa", "corpus")["corpus"]
queries = load_dataset("BeIR/fiqa", "queries")["queries"]

### Dataset samples

Let's see some examples of of the [Financial Opinion Mining and Question Answering (FiQA) dataset](https://huggingface.co/datasets/BeIR/fiqa) documents and queries.

In [ ]:
import pandas as pd
from tabulate import tabulate

console.rule("Corpus Sample")
print(tabulate( 
    corpus
    .to_pandas()
    .head(10)
    .assign(text_start=lambda x: x['text'].str[:100])
    .drop(columns=['text','title'])
    ,headers='keys', 
    tablefmt='github', 
    showindex=False
))


In [ ]:
console.rule("Queries Sample")
print(tabulate( 
    queries
    .to_pandas()
    .head(10)
    .assign(text_start=lambda x: x['text'].str[:100])
    .drop(columns=['text','title'])
    ,headers='keys', 
    tablefmt='github',
    showindex=False
))

## Stage 1: Gather dataset embeddings

CDE works by first getting a set of embeddings from corpus documents that is intended to be representative of the overall corpus. We first sample a number of documents from the corpus (this model is trained with 512 documents from each context) and get their embeddings from our first-stage model.

In [ ]:
query_prefix = "search_query: "
document_prefix = "search_document: "

In [ ]:
import random

def process_ex_document(ex: dict) -> dict:
  ex["text"] = f"{ex['title']} {ex['text']}"
  return ex

corpus_size = improved_model.config.transductive_corpus_size
console.print(f"Choosing {corpus_size} out of {len(corpus)} documents")
minicorpus_docs = corpus.select(random.choices(list(range(len(corpus))), k=corpus_size))
minicorpus_docs = minicorpus_docs.map(process_ex_document)["text"]
minicorpus_docs = tokenizer(
    [document_prefix + doc for doc in minicorpus_docs],
    truncation=True,
    padding=True,
    max_length=512,
    return_tensors="pt"
)

In [ ]:
import torch
device = torch.device("mps" if torch.cuda.is_available() else "cpu")
model.to(device)
minicorpus_docs = minicorpus_docs.to(device)

In [ ]:
import torch
from tqdm.autonotebook import tqdm

batch_size = 32

dataset_embeddings = []
for i in tqdm(range(0, len(minicorpus_docs["input_ids"]), batch_size)):
    minicorpus_docs_batch = {k: v[i:i+batch_size] for k,v in minicorpus_docs.items()}
    with torch.no_grad():
        dataset_embeddings.append(
            improved_model.first_stage_model(**minicorpus_docs_batch)
        )

dataset_embeddings = torch.cat(dataset_embeddings)

## Stage 2: Embed in context

Now that we have dataset embeddings, we can use them to embed queries and documents like normal. We just need to provide an extra argument (`dataset_embeddings` in the CDE code).

In [ ]:
sample_docs = corpus.select(range(16)).map(process_ex_document)["text"]

docs_tokens = tokenizer(
    [document_prefix + doc for doc in sample_docs],
    truncation=True,
    padding=True,
    max_length=512,
    return_tensors="pt"
).to(device)

with torch.no_grad():
  doc_embeddings = improved_model.second_stage_model(
      input_ids=docs_tokens["input_ids"],
      attention_mask=docs_tokens["attention_mask"],
      dataset_embeddings=dataset_embeddings,
  )
doc_embeddings /= doc_embeddings.norm(p=2, dim=1, keepdim=True)

In [ ]:
queries_sample = queries.select(range(16))["text"]
queries_tokens = tokenizer(
    [query_prefix + query for query in queries_sample],
    truncation=True,
    padding=True,
    max_length=512,
    return_tensors="pt"
).to(device)

with torch.no_grad():
  query_embeddings = improved_model.second_stage_model(
      input_ids=queries_tokens["input_ids"],
      attention_mask=queries_tokens["attention_mask"],
      dataset_embeddings=dataset_embeddings,
  )
query_embeddings /= query_embeddings.norm(p=2, dim=1, keepdim=True)

### Models Comparison

Let's compare the two models (basic and improved with context) on the samples of the documents and queries.

In [ ]:
with torch.no_grad():
  doc_basic_embeddings = model.encode(sample_docs)


In [ ]:
with torch.no_grad():
  queries_basic_embeddings = model.encode(queries_sample)

In [ ]:
import matplotlib
import matplotlib.pyplot as plt
import seaborn as sns

fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(20, 8))

# Heatmap for improved model
sns.heatmap((doc_embeddings @ query_embeddings.T).cpu(), cmap="jet", ax=ax1, vmin=0, vmax=1)
ax1.set_title("Improved Model", fontsize=16)

# Heatmap for basic model
sns.heatmap((doc_basic_embeddings @ queries_basic_embeddings.T), cmap="jet", ax=ax2 ,vmin=0, vmax=1)
ax2.set_title("Basic Model", fontsize=16)

plt.tight_layout()
console.rule("Embedding Model Comparison")
plt.show()